core/base/v7/inc/ROOT/RError.hxx

```diff
@@ -0,0 +1,230 @@
+class RResult : public std::conditional_t<std::is_void<T>::value, Internal::RResultTypeVoid, Internal::RResultType<T>> {
+private:
+   /// This is the nullptr for an RResult representing success
+   std::unique_ptr<RError> fError;
+   /// This is safe because checking an RResult is not a multi-threaded operation
+   mutable bool fIsChecked{false};
+
+public:
+   /// Constructor is _not_ explicit in order to allow for `return T();` for functions returning RResult<T>
+   /// Only available if T is not void
+   template <typename Dummy = T, typename = typename std::enable_if_t<!std::is_void<T>::value, Dummy>>
+   RResult(const Dummy &value) : Internal::RResultType<T>(value) {}
+   /// Constructor is _not_ explicit such that the RError returned by R__FAIL can be converted into an RResult<T>
+   /// for any T
+   RResult(const RError &error) : fError(std::make_unique<RError>(error)) {}
+
+   /// RResult<void> has a default contructor that creates an object representing success
+   template <typename Dummy = T, typename = typename std::enable_if_t<std::is_void<T>::value, Dummy>>
+   RResult() {}
+
+   RResult(const RResult &other) = delete;
+   RResult(RResult &&other) = default;
+   RResult &operator =(const RResult &other) = delete;
+   RResult &operator =(RResult &&other) = default;
+
+   ~RResult() noexcept(false)
+   {
+      if (R__unlikely(fError && !fIsChecked)) {
+         // Prevent from throwing if the object is deconstructed in the course of stack unwinding for another exception
+#if __cplusplus >= 201703L
+         if (std::uncaught_exceptions() == 0)
+#else
+         if (!std::uncaught_exception())
+#endif
+         {
+            throw RException(*fError);
+         }
+      }
+   }
+
+   /// Only available if T is not void
+   template <typename Dummy = T>
+   typename std::enable_if_t<!std::is_void<T>::value, const Dummy &>
+   Get()
+   {
+      if (R__unlikely(fError)) {
+         fError->AppendMessage(" (invalid access)");
+         throw RException(*fError);
+      }
+      return Internal::RResultType<T>::fValue;
+   }
+
+   explicit operator bool() const
+   {
+      fIsChecked = true;
+      return !fError;
+   }
+
+   RError *GetError() { return fError.get(); }
+
+   void Throw() { throw RException(*fError); }
+
+   // Help to prevent heap construction of RResult objects. Unchecked RResult objects in failure state should throw
+   // an exception close to the error location. For stack allocated RResult objects, an exception is thrown
+   // the latest when leaving the scope. Heap allocated ill RResult objects can live much longer making it difficult
+   // to trace back the original failure.
+   void *operator new(std::size_t size) = delete;
+   void *operator new(std::size_t, void *) = delete;
+   void *operator new[](std::size_t) = delete;
+   void *operator new[](std::size_t, void *) = delete;
+};
+
+/// Short-hand to return an RResult<void> indicating success
+#define R__SUCCESS return ROOT::Experimental::RResult<void>();
```

Comment:
If it cannot go away, `return` should not be a part of it, such that an invocation reads:
```
return R__SUCCESS;
```
(But I'd really prefer to use a static constant.)